### ISSUE TITLE
[ENHANCEMENT] speed up `Hantush*.step()`

### ISSUE THREAD

**Issue by dbrakenhoff** [pastas/pastas#430]
**Describe the proposed enhancement**
I've been running into (extremely) slow model solve times using `ps.WellModel`, especially when using `ps.LmFitSolve`. 

Occasionally, the small value of parameter _b_ causes the optimization to run for a long time. Reformulating the implementation so that _b_ = `np.exp(b_new)` seems to take care of this in some cases, and rids us of those extremely small values for _b_. 

I also came across https://github.com/numba/numba-scipy, which when installed allows numba acceleration of `scipy.special` functions. I rewrote `HantushWellModel.step()` with a numba implementation (see below). This gives a 2x speedup. The downside is that this would add an optional dependency, and currently `numba-scipy` only supports `scipy<=1.7.3` (current version is 1.9.1). There has been a [recent discussion](https://github.com/numba/numba-scipy/issues/88) on upgrading the dependency once scipy 1.9.1 is available through anaconda. It would perhaps be nice to include a `numba=True` option in HantushWellModel and Hantush to make this possible.

```python
class HantushWellModel(RfuncBase):

    # ...

    @staticmethod
    @njit(parallel=True)
    def numba_step(A, a, b, r, t):
        rho = 2 * r * np.exp(b / 2)
        rhosq = rho**2
        k0rho = k0(rho)
        tau = t / a
        w = (exp1(rho) - k0rho) / (exp1(rho) - exp1(rho / 2))
        F = np.zeros((tau.size,), dtype=np.float64)
        for i in prange(tau.size):
            tau_i = tau[i]
            if tau_i < rho / 2:
                F[i] = w * exp1(rhosq / (4 * tau_i)) - (w - 1) * exp1(
            tau_i + rhosq / (4 * tau_i))
            elif tau_i >= rho / 2:
                F[i] = 2 * k0rho - w * exp1(tau_i) + (w - 1) * exp1(
            tau_i + rhosq / (4 * tau_i))
        return A * F / 2

    def step(self, p, dt=1, cutoff=None, maxtmax=None):
        r = self._get_distance_from_params(p)
        A, a, b = p[:3]
        t = self.get_t(p, dt, cutoff, maxtmax)
        return self.numba_step(A, a, b, r, t)
```

Timing results for some random model I've been working with:
```
Optimization progress: 1408it [02:18, 10.16it/s]  # current implementation
Optimization progress: 1258it [00:53, 23.66it/s]  # numba implementation
```
So over 2x faster, not amazing, but certainly nice.

**Comment by dbrakenhoff** [comment]
I did some testing and the numba implementation is faster only when parameter _a_ is bigger than 100 (probably a bit less), but then it is much faster given the longer run times to calculate `step()`. 

Here the results for some timeit runs for different combinations of _a_ and _b_. 

![image](https://user-images.githubusercontent.com/6552225/194517660-b3e1383b-b1cf-4caa-8e0f-9cbf6dc61218.png)

**Comment by HenrikSpa** [comment]
This implementation would be great!
I came to the same conclusion and the same solution, so that b_new = np.log(b_old), and your suggestion of b = np.exp(b) for the step calculation. This problem happend especially for large distances (my current model has distances from 1 to 1000 meters).

I also did the same with the A parameter since it also can get rather small (much smaller than the regular Hantush A parameter) but then the sign has to be changed and rechanged again for calculation as well as you can't take np.log() of a negative number.

**Comment by HenrikSpa** [comment]
Maybe you also have to make a change in get_tmax where b is used?:

https://github.com/pastas/pastas/blob/1cd9fd94502920de8ae1bb03a902fd0d5ab25354/pastas/rfunc.py#L346

**Comment by raoulcollenteur** [comment]
@dbrakenhoff please check if we can close this issue after I merged the PR #430 just now.
